Wikipedia/Code/SavedArticlesFetcher.m

```diff
@@ -163,18 +179,91 @@ - (void)fetchTitle:(MWKTitle*)title failure:(WMFErrorHandler)failure success:(WM
 }
 
 - (void)downloadImageDataForArticle:(MWKArticle*)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
+    if (![[NSUserDefaults standardUserDefaults] wmf_didFinishLegacySavedArticleImageMigration]) {
+        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0.5)
+        [self migrateLegacyImagesInArticle:article];
+    }
     [self fetchAllImagesInArticle:article failure:^(NSError *error) {
         failure([NSError wmf_savedPageImageDownloadError]);
     } success:^{
         [self fetchGalleryDataForArticle:article failure:failure success:success];
     }];
 }
 
+- (void)migrateLegacyImagesInArticle:(MWKArticle *)article {
+    //  Removes up old saved article image list folders, copies old cached images to original and article image width locations. This ensures articles saved with 5.0.4 and older will still have images availble offline in 5.0.5. The migration is idempotent - the enumerated folders are removed so they won't be processed the next time around.
+    
+    //Get the folder that contains legacy saved article images for example - articles/Barack_Obama/Images/
+    NSString *imagesFolderPath = [self.savedPageList.dataStore pathForImagesWithTitle:article.title];
+    NSURL *imagesFolderURL = [NSURL fileURLWithPath:imagesFolderPath isDirectory:YES];
+    NSEnumerator *enumerator = [[NSFileManager defaultManager] enumeratorAtURL:imagesFolderURL includingPropertiesForKeys:@[NSURLIsDirectoryKey] options:NSDirectoryEnumerationSkipsSubdirectoryDescendants errorHandler:^BOOL(NSURL * _Nonnull url, NSError * _Nonnull error) {
+        DDLogError(@"Error enumerating image directory: %@", error);
+        return YES;
+    }];
+    
+    WMFImageController *imageController = [WMFImageController sharedInstance];
+    NSUInteger articleImageWidth = [[UIScreen mainScreen] wmf_articleImageWidthForScale];
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    
+    for (NSURL *imageFolderURL in enumerator) { //Enumerate each subfolder of the image folder. There is one subfolder per saved image.
```

Comment:
No need to walk the directory structure opening an Image.plist for each image... I think there's an index of all the image urls in Image**s**.plist:

![screen shot 2016-07-21 at 5 41 36 pm](https://cloud.githubusercontent.com/assets/3143487/17043312/6e6dddd6-4f6a-11e6-943a-881f6523ecad.png)